### PR TITLE
Menu API

### PR DESCRIPTION
**Pull request by nickrobson**
_TL;DR: Adds an API for creating multi-layer menus._

**There is currently no menu API, and this PR hopes to fill this gap.**

**Example Usage**

```java
MenuLoader.registerListeners(bot);

final Menu menu = MenuLoader.createMenu(bot, "Loading...");
final MenuScreen firstScreen = menu.createScreen(UUID.nameUUIDFromBytes("First Screen".getBytes()), () -> "First Screen", ParseMode.NONE);
final MenuScreen secondScreen = menu.createScreen(UUID.nameUUIDFromBytes("Second Screen".getBytes()), () -> "Second Screen", ParseMode.NONE);

menu.addContextLoadListener(bm -> {
    System.out.println("Loading MenuContext!");
});

firstScreen.getGrid().addRow(
        menu.createButton(b -> b.label("Go to Second Screen"), (m, u) -> {
            m.setScreen(secondScreen);
            return null;
        })
);
secondScreen.getGrid().addRow(
        menu.createButton(b -> b.label("Go to First Screen"), (m, u) -> {
            m.setScreen(firstScreen);
            return null;
        })
);

bot.getEventRegistry().registerEvent(TextMessageEvent.class, e -> {
    try {
        BoundMenu boundMenu = menu.send(e.getBot(), e.getMessage().getChat(), e.getMessage().getMessageId());
        boundMenu.setScreen(firstScreen);
        boundMenu.update();
    } catch (TelegramException ex) {
        ex.printStackTrace();
    }
});
```

---

**Motivation**

One thing I found lacking in @zackpollard's Telegram Bot API for Java was a good, easy-to-use, stateless menu API... one that could very easily be made to be multiple layers deep if wanted. (Sorry, Mazen.)

To solve this, I ended up creating my own menu API, [telepad](https://github.com/nickrobson/telepad).
Despite my personal bias, I found this (telepad) API to be far cleaner than the provided one in Zack's API.

This PR aims to be similar in some ways to the telepad menu API, and in other ways more easily understood and extended. It very much aims to put the power in the developer's hands with what is possible, providing pretty much all the data the bot has to the developer to use in whatever way they want — the API just provides a simple way of linking together multiple callback queries!

Sending the same menu to multiple different chats works, albeit currently they share state — clicking on a button on one menu will "shadow-change" the state on the other (it won't auto-update the other menu as messages sent are not stored, but the other menu will not allow interaction until it has been updated, which it does automatically).

### CODE REVIEW

File: jtelegrambotapi-core/src/main/java/com/jtelegram/api/ex/TelegramException.java [hunk 1]
```diff
@@ -2,10 +2,17 @@
...
 
 @Getter
+@ToString
 public class TelegramException extends Exception {
     private int errorCode;
     private String description;
     private ResponseParameters parameters;
+
+    @Override
+    public String getMessage() {
```
Comment (aaomidi) [inline jtelegrambotapi-core/src/main/java/com/jtelegram/api/ex/TelegramException.java]: This is a good change.

File: jtelegrambotapi-menus/src/main/java/com/jtelegram/api/menu/MenuStateMemory.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+    /**
+     * Removes the current state, and sets the previous state as the new state.
+     *
+     * @return The old state.
+     *
+     * @throws StateMemoryDepletedException If there are no previous states.
+     */
+    @Nonnull
+    MenuState popState() throws StateMemoryDepletedException;
```
Comment (mkotb) [inline jtelegrambotapi-menus/src/main/java/com/jtelegram/api/menu/MenuStateMemory.java]: Popping a state shouldn't throw an exception. If there are no previous states, return null. Typed exceptions are:
a) slower than they need to be
b) an annoyance to deal with

So unless this problem will be extremely common or there will be a variety of errors that arise from the operation (think IO ops and whatnot), it's better off to find a way to not throw the exception but still allow the developer to deal with the error.

File: jtelegrambotapi-menus/src/main/java/com/jtelegram/api/menu/impl/MenuImpl.java [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+public class MenuImpl implements Menu {
+
+    private static final AtomicInteger menuIdCounter = new AtomicInteger(0);
```
Comment (mkotb) [inline jtelegrambotapi-menus/src/main/java/com/jtelegram/api/menu/impl/MenuImpl.java]: Menu IDs here are generated on the fly, increment from zero, but the result of this is that when the bot restarts, old menus will make execute callbacks to newly generated ones as they share the same id from different states.

The solution for this is to create menus with string identifiers provided by the developer (A-z,0-9), and allow menus to have multiple viewers (which I think you allow here). So instead of your callback data being `menuId stateId row button` it should be `menuId chatId stateId row button`. An additional consideration would be how you allow developers to add metadata to the `MenuContext` class, since that isn't persistent either. Removing this ability would force the developer to implement their own chat -> context appropriately for the menu/bot they are building, which would be more functional than forgetting everything on restart.

If you want, you could implement a saving system for `MenuContext` and take care of it for the developer, providing them with different options, but I think that goes beyond the reach of this API.

File: jtelegrambotapi-menus/src/main/java/com/jtelegram/api/menu/impl/BoundMenuImpl.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    @Nonnull
+    @Override
+    public TextMessage getMessage() {
+        return this.message;
+    }
+
+    @Override
+    public void update(@Nonnull TelegramBot bot) {
+        if (message.getSender().getId() != bot.getBotInfo().getId()) {
+            throw new IllegalArgumentException("Message was not sent by me!");
```
Comment (mkotb) [inline jtelegrambotapi-menus/src/main/java/com/jtelegram/api/menu/impl/BoundMenuImpl.java]: This should actually specify which bot it wasn't sent from as this API supports multiple bots operating concurrently

File: jtelegrambotapi-menus/src/main/java/com/jtelegram/api/menu/impl/MenuImpl.java [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        try {
+            countDownLatch.await();
+        } catch (InterruptedException ex) {
+            throw new RuntimeException(ex);
+        }
+        Object moe = messageOrError.get();
+        if (moe instanceof TextMessage) {
+            TextMessage message = (TextMessage) moe;
+            BoundMenuImpl boundMenu = new BoundMenuImpl(this, message);
+            menusByMessageId.put(message.getMessageId(), boundMenu);
```
Comment (mkotb) [inline jtelegrambotapi-menus/src/main/java/com/jtelegram/api/menu/impl/MenuImpl.java]: Okay, so at first I thought this was a really smart idea until I remembered that message ids are not unique across different chats. IIRC They increment from a starting point and two messages can share the same message id given that they are from two different chats. Another solution is to add the chat id on top of that but... at this point I'd suggest, like before, having ids specified by the developer that uniquely identify the menu. To avoid any weird persistence issues, this would mean that the metadata thing you did here would have to be removed. If the IDs are specified by the developer, they can handle persistence across restarts.

IMO a developer shouldn't have to worry that restarting their bot will break their end users' experiences. It should be seamless and this system should support it.

File: jtelegrambotapi-menus/src/main/java/com/jtelegram/api/menu/BackButton.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+import com.jtelegram.api.user.User;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/**
+ * Goes back to the previous menu.
+ */
+public class BackButton implements MenuButton {
+
+    @Nonnull
```
Comment (mkotb) [inline jtelegrambotapi-menus/src/main/java/com/jtelegram/api/menu/BackButton.java]: Do IDEs such as IntelliJ support this annotation for NPE warnings?

File: jtelegrambotapi-menus/src/main/java/com/jtelegram/api/menu/impl/MenuImpl.java (hunk 5 above)
Comment (nickrobson) [reply jtelegrambotapi-menus/src/main/java/com/jtelegram/api/menu/impl/MenuImpl.java]: RIP I forgot about the lack of uniqueness.

File: jtelegrambotapi-menus/src/main/java/com/jtelegram/api/menu/BackButton.java (hunk 6 above)
Comment (nickrobson) [reply jtelegrambotapi-menus/src/main/java/com/jtelegram/api/menu/BackButton.java]: Yup! It actually does a great job.

**Comment by nickrobson** [comment]
I've made it use the chatId and the messageId in combination. You should have push access to my branch, so if you want to make the ID developer-configured, please go ahead. I personally don't have the motivation to do that specifically.

**Comment by nickrobson** [comment]
Regarding JSR305, IntelliJ has inbuilt support for the annotations, and provides tooltips if it believes something will be in breach of them. JSR305 was going to be accepted into Java 7/8 but Oracle keeps delaying it, and the Google implementation is a working reference implementation.

File: jtelegrambotapi-menus/src/main/java/com/jtelegram/api/menu/impl/MenuImpl.java [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+            return; // invalid format, best not to try
+        }
+        // shouldn't happen
+        MenuScreenImpl screen = boundMenu.getScreen();
+        if (screen.screenId != screenId) {
+            // user clicking an outdated menu
+            // tell them, and update it
+            bot.perform(AnswerCallbackQuery.builder()
+                    .queryId(event.getQuery().getId())
+                    .text("The menu you clicked on is out of date!\nI'm updating it now for you.")
```
Comment (aaomidi) [inline jtelegrambotapi-menus/src/main/java/com/jtelegram/api/menu/impl/MenuImpl.java]: You shouldn't use English, even if it's never going to happen.

**Comment by aaomidi** [comment]
Alright, so this looks good. The @Nonnull stuffs are a great addition.

I want to ask you about a slightly different version of what you're doing.

1. Can we start using UUIDs instead of AtomicIntegers/Integers?
2. If we do that, can we let the developer get the UUID of the Menu they're making/made?
3. Can we allow the developer to register known menus upon startup of the bot? This would essentially be, "hey if its UUID X, call my method Y". This would be done by the developer storing a tuple of chatID-UUID, and remembering the "state" it is in.
4. For the lazy devs who just don't want to serialize, can we have a "default" handler of unregistered menus that users click on? Essentially if you don't recognize UUID X, run this method on it. The method could simply edit the message with something like "Bot restarted - do /start" or something like that - up to the user.

On a side note, avoid adding any specific strings to the api that the developer can't override based on their language.

**Comment by nickrobson** [comment]
@aaomidi Thanks for the review.

To attempt to answer the questions:
1. Yes, I definitely agree that this would be a worthwhile change. It would only require `MenuScreen` to be able to take one, and then the developer can handle stuff there. Note that a `MenuGrid` is contained inside a menu screen, and it's only there that this ID matters.
2. Since a `Menu` is just the entrypoint to the API, and stores no state, it doesn't make sense for it to use a UUID. You might as well store the `Menu` object itself and use that.
3. Yes, being able to specify the UUID for a `MenuScreen` is a great idea. Perhaps we can enforce it uses a type 4 (random) UUID by default and when you set it, it _needs_ to be a type 3 (user-set) UUID? Then as you said, they can just store a `(chatId, messageId, menuScreenUUID)` tuple so they know what menu the user is on. Storing state across this would of course need them to use a DB solution, but I believe that's far out of the scope of this API.
4. Yeah, sure. That can be done. That'd be called when there's no menu set to that chat/message ID combination, I suppose.

Regarding the string, I know. I'll make it settable, defaulting to English – which I believe is a reasonable default.

**Comment by aaomidi** [comment]
@nickrobson All of these sound good. Sorry for the confusion about menus.

**Comment by nickrobson** [comment]
@aaomidi @mkotb 

I've made the requested changes.
`MenuScreen`s can be provided with a UUID as the first argument to `Menu#createScreen` (see example). When this is done, they will persist across restarts. I've added the ability to load data into the `MenuContext` class when it is first created, to allow for data recovery. This way updating data can be the only one requiring DB access (and setting into the context object), and gets can just use the context object. 😄 

Let me know what you think.
